### PR TITLE
Improvement/arm support1

### PR DESCRIPTION
**Pull request by weizhou-chaojixx**
Add ARM Cortex-m KVM option on X86.

### CODE REVIEW

File: roms/vgabios/config.mak [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Automatically generated by configure - do not modify
```
Comment (vitalych) [inline roms/vgabios/config.mak:1]: Does this have to be committed?

File: .gitignore [hunk 2]
```diff
@@ -1,3 +1,5 @@
+my.vim*
+*.swp
```
Comment (vitalych) [inline .gitignore:2]: This should go to a separate commit

File: roms/vgabios/config.mak [hunk 3]
```diff
@@ -1,5 +1,5 @@
 # Automatically generated by configure - do not modify
-SRC_PATH=/opt/qemu-1/roms/vgabios
+SRC_PATH=/opt/s2earm/source/s2e/qemu/roms/vgabios
```
Comment (vitalych) [inline roms/vgabios/config.mak:2]: Please squash all commits, then clean up files that should not be modified, then redo new commits, logically separated.

File: target/arm/kvm32.c [hunk 4]
```diff
@@ -173,51 +173,58 @@ int kvm_arch_init_vcpu(CPUState *cs)
     struct kvm_one_reg r;
     ARMCPU *cpu = ARM_CPU(cs);
 
-    if (cpu->kvm_target == QEMU_KVM_ARM_TARGET_NONE) {
-        fprintf(stderr, "KVM is not supported for this guest CPU type\n");
-        return -EINVAL;
+    if (arm_feature(&cpu->env, ARM_FEATURE_M)){
+        ret = kvm_cortex_m_vcpu_init(cs);
```
Comment (vitalych) [inline target/arm/kvm32.c]: just do return here, to avoid having the else branch.

Comment (vitalych) [reply target/arm/kvm32.c]: Missing space before {. Unfortunately, we don't have clang-format for qemu yet.

File: accel/kvm/kvm-all.c [hunk 5]
```diff
@@ -2265,12 +2265,19 @@ int kvm_cpu_exec(CPUState *cpu)
...
         case KVM_EXIT_MMIO:
             DPRINTF("handle_mmio\n");
-            /* Called outside BQL */
-            address_space_rw(&address_space_memory,
-                             run->mmio.phys_addr, attrs,
-                             run->mmio.data,
-                             run->mmio.len,
-                             run->mmio.is_write);
+            /* SCS region in armv7m space memory */
+            if(run->mmio.phys_addr>0xe0000000)
```
Comment (vitalych) [inline accel/kvm/kvm-all.c:2269]: Check code style, use { } in all blocks and spaces after if.

File: exec.c [hunk 6]
```diff
@@ -898,20 +900,21 @@ void cpu_address_space_init(CPUState *cpu, int asidx,
                             const char *prefix, MemoryRegion *mr)
 {
     CPUAddressSpace *newas;
-    AddressSpace *as = g_new0(AddressSpace, 1);
+   // AddressSpace *as = g_new0(AddressSpace, 1);
```
Comment (vitalych) [inline exec.c:903]: Remove all commented out code

File: target/arm/cpu.h [hunk 7]
```diff
@@ -1698,7 +1699,7 @@ void armv7m_nvic_acknowledge_irq(void *opaque);
  *           0 if there is still an irq active after this one was completed
  * (Ignoring -1, this is the same as the RETTOBASE value before completion.)
  */
-int armv7m_nvic_complete_irq(void *opaque, int irq, bool secure);
+//int armv7m_nvic_complete_irq(void *opaque, int irq, bool secure);
```
Comment (vitalych) [inline target/arm/cpu.h:1702]: Why were these commented out?

File: target/arm/kvm32.c [hunk 8]
```diff
@@ -313,83 +313,88 @@ int kvm_arch_put_registers(CPUState *cs, int level)
...
-        memcpy(env->fiq_regs, env->regs + 8, 5 * sizeof(uint32_t));
-    } else {
-        memcpy(env->usr_regs, env->regs + 8, 5 * sizeof(uint32_t));
+    if (arm_feature(&cpu->env, ARM_FEATURE_M)){
+        return kvm_cortex_m_vcpu_init(cs);
     }
-    env->banked_r13[bn] = env->regs[13];
-    env->banked_r14[bn] = env->regs[14];
-    env->banked_spsr[bn] = env->spsr;
+    else{
```
Comment (vitalych) [inline target/arm/kvm32.c:319]: No need for else after a return.

**Comment by vitalych** [comment]
You don't need to close/delete the pull request. Just force-push your updates. It makes the job of reviewers much simpler. I can quickly check if comments have been addressed that way.

**Comment by weizhou-chaojixx** [comment]
I have completed reorganized the original commits and use a new clear
branch, you can just delete the original. The new PR is more simple and
clear. I think you can do a quick check.

Vitaly Chipounov <notifications@github.com> 于2019年10月15日周二 下午3:45写道：

**Comment by vitalych** [comment]
Instead of just doing a pull, reviewers will have to checkout new branches across several repos if they need to test them. Force-pushing avoid this problem. Regarding reviews, github is smart enough to keep the previous comments around, and you can click on "resolve" underneath each comment to keep track of those that you addressed. It also keeps all the previous discussions. All of that is lost if you create a new PR.

I suppose you copied your code to a new branch and committed from scratch. You can avoid this by using git rebase in interactive mode (git rebase -i), which lets you edit individual commits / squash / reword commit messages, etc. If you mess something up, you can use "git rebase --abort" or  "git reflog" to go back to the last good revision.

**Comment by weizhou-chaojixx** [comment]
Since I just delete the original PR branch, I just checkout a new branch
and rebase it in interactive mode. Sorry, next time I will directly edit
the original PR. That will only happen at this time. By the way, if I
rebase a commit (e.g., I split a commit into two), the github can still
keep the previous comments around?

Vitaly Chipounov <notifications@github.com> 于2019年10月16日周三 上午5:27写道：
